File: Android_integration_tests/Firebase_java_empty_activity/app/src/main/java/com/myfirstapp/firebase_java_test/MainActivity.java
package com.myfirstapp.firebase_java_test;

import static androidx.core.content.ContentProviderCompat.requireContext;

import static java.security.AccessController.getContext;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.myfirstapp.firebase_java_test.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    ActivityMainBinding binding;
    String CelestialBody;
    FirebaseDatabase db;
    DatabaseReference reference;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        String[] celestial = getResources().getStringArray(R.array.celestial);
        ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(this, R.layout.dropdown_item,celestial);
        binding.autoCompleteTextView.setAdapter(arrayAdapter);



        binding.button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                CelestialBody = binding.autoCompleteTextView.getText().toString();

                if(!CelestialBody.isEmpty()){
                    Celestial celestial = new Celestial(CelestialBody);
                    db = FirebaseDatabase.getInstance();
                    reference = db.getReference("Planet");
                    reference.setValue(CelestialBody).addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {

                            binding.autoCompleteTextView.setText("");
                            Toast.makeText(MainActivity.this, "Succesfully updated", Toast.LENGTH_SHORT).show();

                        }
                    });
                }

            }
        });
    }
}

File: Codes/App_Development/app/src/main/java/com/earendel/earendel/MainActivity.kt
package com.earendel.earendel

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.widget.Button
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.firebase.database.FirebaseDatabase

class MainActivity : AppCompatActivity() {

    private val databaseReference = FirebaseDatabase.getInstance().reference

    private lateinit var fusedLocationProviderClient: FusedLocationProviderClient

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        fusedLocationProviderClient=LocationServices.getFusedLocationProviderClient(this)

        getCurrentLocation();

            val starButton = findViewById<Button>(R.id.btnStar)
            val planetButton = findViewById<Button>(R.id.btnPlanet)

        starButton.setOnClickListener{
            val intent = Intent(this, StarActivity::class.java)
            startActivity(intent)
        }

        planetButton.setOnClickListener{
            val intent = Intent(this, PlanetActivity::class.java)
            startActivity(intent)
        }

    }

    private fun getCurrentLocation(){
        if(checkPermissions()){
            if(isLocationEnabled()){

                if (ActivityCompat.checkSelfPermission(
                        this,
                        Manifest.permission.ACCESS_FINE_LOCATION
                    ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                        this,
                        Manifest.permission.ACCESS_COARSE_LOCATION
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    requestPermission()
                    return
                }
                fusedLocationProviderClient.lastLocation.addOnCompleteListener(this){ task->
                    val location: Location?=task.result
                    if(location==null){
                        Toast.makeText(this,"Not Received",Toast.LENGTH_SHORT).show()
                    }
                    else{
                        Toast.makeText(this,"Turn on Location",Toast.LENGTH_SHORT).show()
                        //Add to firebase
                        val locationRef = databaseReference.child("Location")
                        val locationMap = HashMap<String, Any>()
                        locationMap["latitude"] = location.latitude
                        locationMap["longitude"] = location.longitude
                        //Log.d("TAG", location)
                        locationRef.setValue(locationMap)
                    }
                }

            }
            else{
                //setting open here
                Toast.makeText(this,"Turn on Location",Toast.LENGTH_SHORT).show()
                val intent=Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
                startActivity(intent)
            }
        }
        else{
            //request permissions here
            requestPermission()
        }
    }

    private fun isLocationEnabled():Boolean{
        val locationManager:LocationManager=getSystemService(Context.LOCATION_SERVICE) as LocationManager
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)||locationManager.isProviderEnabled(
            LocationManager.NETWORK_PROVIDER
        )
    }

    private fun requestPermission() {
        ActivityCompat.requestPermissions(
            this, arrayOf(android.Manifest.permission.ACCESS_COARSE_LOCATION,
            android.Manifest.permission.ACCESS_FINE_LOCATION),
            PERMISSION_REQUEST_ACCESS_LOCATION
        )
    }

    companion object{
        private const val PERMISSION_REQUEST_ACCESS_LOCATION=100
    }

    private fun checkPermissions(): Boolean{
        if(ActivityCompat.checkSelfPermission(this,
            android.Manifest.permission.ACCESS_COARSE_LOCATION)
            ==PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this,
            android.Manifest.permission.ACCESS_FINE_LOCATION)==PackageManager.PERMISSION_GRANTED){

            return true
        }
        return false

    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if(requestCode== PERMISSION_REQUEST_ACCESS_LOCATION){
            if(grantResults.isNotEmpty() && grantResults[0]==PackageManager.PERMISSION_GRANTED){
                Toast.makeText(applicationContext,"Permission Granted",Toast.LENGTH_SHORT).show()
                getCurrentLocation()
            }
            else{
                Toast.makeText(applicationContext,"Permission Denied",Toast.LENGTH_SHORT).show()
            }
        }

    }

}